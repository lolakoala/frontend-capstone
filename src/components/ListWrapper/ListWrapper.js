import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TextField } from 'material-ui';
import BuddyList from '../BuddyList/BuddyList';
import ProfsList from '../ProfsList/ProfsList';

class ListWrapper extends Component {
  constructor() {
    super();
    this.state = {
      value: '',
      list: [],
      listName: ''
    };
  }

  componentWillMount() {
    const { currentUser, location, getBuddies, getPreferredProfs } = this.props;

    if (location.pathname === '/userBuddies') {
      getBuddies(currentUser);
    }
    if (location.pathname === '/userProfs') {
      getPreferredProfs(currentUser);
    }
  }

  componentDidMount() {
    const {
      location,
      buddySearch,
      profSearch,
      userBuddies,
      userProfs
    } = this.props;
    let list;
    let listName;

    switch (location.pathname) {
    case '/userBuddies':
      list = userBuddies;
      listName = 'buddies';
      break;
    case '/userProfs':
      list = userProfs;
      listName = 'preferred professionals';
      break;
    case '/profSearch':
      list = profSearch;
      listName = 'professionals search results';
      break;
    case '/buddySearch':
      list = buddySearch;
      listName = 'buddy search results';
      break;
    default:
      list = [];
    }

    return this.setState({ list, listName });
  }

  handleChange = (event) => {
    this.setState({
      value: event.target.value
    });
  };

  filterList = array => {
    return array.filter(person => {
      const searchablePerson = Object.assign({}, person, { city: '' });
      const personValues = Object.values(searchablePerson);
      let isMatch = false;
      personValues.forEach(value => {
        if (value.includes(this.state.value)) {
          isMatch = true;
        }
      });
      return isMatch === true;
    });
  }

  render() {
    const { location, currentUser, toggleFavorite } = this.props;
    const listToRender = this.state.value ? this.filterList(this.state.list) : this.state.list;
    const buddyComponent = <BuddyList
      buddies={listToRender}
      toggleFavorite={toggleFavorite}
      currentUser={currentUser}
    />;
    const profsComponent = <ProfsList
      profs={listToRender}
      toggleFavorite={toggleFavorite}
      currentUser={currentUser}
    />;

    return (
      <div>
        <TextField
          value={this.state.value}
          onChange={this.handleChange}
          hintText={`Type here to filter your ${this.state.listName}.`}
        />
        {location.pathname.includes('bud') ? buddyComponent : profsComponent}
      </div>
    );
  }
}

export default ListWrapper;

ListWrapper.propTypes = {
  location: PropTypes.object,
  getBuddies: PropTypes.func,
  currentUser: PropTypes.object,
  buddySearch: PropTypes.object,
  getPreferredProfs: PropTypes.func,
  profSearch: PropTypes.array,
  toggleFavorite: PropTypes.func,
  userBuddies: PropTypes.array,
  userProfs: PropTypes.array
};
