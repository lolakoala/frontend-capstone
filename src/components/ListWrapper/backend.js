const backend = 'https://mental-healthy-backend.herokuapp.com';

export default backend;
