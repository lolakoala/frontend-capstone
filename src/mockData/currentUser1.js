const currentUser1 = {
  email: 'lola@gmail.com',
  city: 'Denver',
  userName: 'lolakoala',
  aboutMe: 'Lola loves doggos.'
};

export default currentUser1;
