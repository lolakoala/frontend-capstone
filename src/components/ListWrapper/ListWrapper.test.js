import React from "react";
import ListWrapper from './ListWrapper.js';
import { shallow, configure } from "enzyme";
import currentUser1 from '../../mockData/currentUser1.js';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('ListWrapper', () => {
  const mockFn = jest.fn();
  const wrapper = shallow(<ListWrapper
    location={{ pathname: '/listUserBuddies'}}
    getBuddies={mockFn}
    currentUser={currentUser1}
    buddySearch={[]}
    getPreferredProfs={mockFn}
    profSearch={[]}
    toggleFavorite={mockFn}
    userBuddies={[currentUser1]}
    userProfs={[]}
  />);

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should match snap', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set state on mount', () => {
    expect(wrapper.state()).toEqual({
      value: '',
      list: [currentUser1],
      listName: 'buddies'
    });
  });

  it('should call 2 functions on mount', () => {
    expect(mockFn).toHaveBeenCalledTimes(2);
  });

  it('should have textfield', () => {
    const search = wrapper.find('TextField');

    expect(search.length).toEqual(1);
  });

  it('should change state on change of inputs', () => {
    const input = wrapper.find('TextField');

    input.simulate('change', { target: { value: 'lola' } });
    expect(wrapper.state().value).toEqual('lola');
  });

  it('should render BuddyList', () => {
    const budList = wrapper.find('BuddyList');

    expect(budList.length).toEqual(1);
  });

  it('should render profList is pathname does not include bud', () => {
    const wrapper2 = shallow(<ListWrapper
      location={{ pathname: '/listUserProfs'}}
      getBuddies={mockFn}
      currentUser={currentUser1}
      buddySearch={[]}
      getPreferredProfs={mockFn}
      profSearch={[]}
      toggleFavorite={mockFn}
      userBuddies={[currentUser1]}
      userProfs={[]}
    />);
    const profList = wrapper2.find('ProfsList');

    expect(profList.length).toEqual(1);
  });
});
