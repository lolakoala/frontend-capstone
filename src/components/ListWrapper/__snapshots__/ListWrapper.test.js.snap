// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListWrapper should match snap 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ListWrapper
    buddySearch={Array []}
    currentUser={
        Object {
            "aboutMe": "Lola loves doggos.",
            "city": "Denver",
            "email": "lola@gmail.com",
            "userName": "lolakoala",
          }
    }
    getBuddies={[Function]}
    getPreferredProfs={[Function]}
    location={
        Object {
            "pathname": "/list/userBuddies",
          }
    }
    profSearch={Array []}
    toggleFavorite={[Function]}
    userBuddies={
        Array [
            Object {
              "aboutMe": "Lola loves doggos.",
              "city": "Denver",
              "email": "lola@gmail.com",
              "userName": "lolakoala",
            },
          ]
    }
    userProfs={Array []}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          fullWidth={false}
          hintText="Type here to filter your buddies."
          multiLine={false}
          onChange={[Function]}
          rows={1}
          type="text"
          underlineShow={true}
          value=""
/>,
        <BuddyList
          buddies={
                    Array [
                              Object {
                                "aboutMe": "Lola loves doggos.",
                                "city": "Denver",
                                "email": "lola@gmail.com",
                                "userName": "lolakoala",
                              },
                            ]
          }
          currentUser={
                    Object {
                              "aboutMe": "Lola loves doggos.",
                              "city": "Denver",
                              "email": "lola@gmail.com",
                              "userName": "lolakoala",
                            }
          }
          faves={
                    Array [
                              Object {
                                "aboutMe": "Lola loves doggos.",
                                "city": "Denver",
                                "email": "lola@gmail.com",
                                "userName": "lolakoala",
                              },
                            ]
          }
          toggleFavorite={[Function]}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "floatingLabelFixed": false,
          "fullWidth": false,
          "hintText": "Type here to filter your buddies.",
          "multiLine": false,
          "onChange": [Function],
          "rows": 1,
          "type": "text",
          "underlineShow": true,
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "buddies": Array [
            Object {
              "aboutMe": "Lola loves doggos.",
              "city": "Denver",
              "email": "lola@gmail.com",
              "userName": "lolakoala",
            },
          ],
          "currentUser": Object {
            "aboutMe": "Lola loves doggos.",
            "city": "Denver",
            "email": "lola@gmail.com",
            "userName": "lolakoala",
          },
          "faves": Array [
            Object {
              "aboutMe": "Lola loves doggos.",
              "city": "Denver",
              "email": "lola@gmail.com",
              "userName": "lolakoala",
            },
          ],
          "toggleFavorite": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <TextField
            disabled={false}
            floatingLabelFixed={false}
            fullWidth={false}
            hintText="Type here to filter your buddies."
            multiLine={false}
            onChange={[Function]}
            rows={1}
            type="text"
            underlineShow={true}
            value=""
/>,
          <BuddyList
            buddies={
                        Array [
                                    Object {
                                      "aboutMe": "Lola loves doggos.",
                                      "city": "Denver",
                                      "email": "lola@gmail.com",
                                      "userName": "lolakoala",
                                    },
                                  ]
            }
            currentUser={
                        Object {
                                    "aboutMe": "Lola loves doggos.",
                                    "city": "Denver",
                                    "email": "lola@gmail.com",
                                    "userName": "lolakoala",
                                  }
            }
            faves={
                        Array [
                                    Object {
                                      "aboutMe": "Lola loves doggos.",
                                      "city": "Denver",
                                      "email": "lola@gmail.com",
                                      "userName": "lolakoala",
                                    },
                                  ]
            }
            toggleFavorite={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "floatingLabelFixed": false,
            "fullWidth": false,
            "hintText": "Type here to filter your buddies.",
            "multiLine": false,
            "onChange": [Function],
            "rows": 1,
            "type": "text",
            "underlineShow": true,
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "buddies": Array [
              Object {
                "aboutMe": "Lola loves doggos.",
                "city": "Denver",
                "email": "lola@gmail.com",
                "userName": "lolakoala",
              },
            ],
            "currentUser": Object {
              "aboutMe": "Lola loves doggos.",
              "city": "Denver",
              "email": "lola@gmail.com",
              "userName": "lolakoala",
            },
            "faves": Array [
              Object {
                "aboutMe": "Lola loves doggos.",
                "city": "Denver",
                "email": "lola@gmail.com",
                "userName": "lolakoala",
              },
            ],
            "toggleFavorite": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
